Mapping hazards and types

Thera-Pissed opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
## About The Pull Request
## Baseline functionality finished! yay
(still plenty of room for actual hazards, mostly need sprites & ideas)
### **Pitch**
This PR implements a base obj/structure/hazard, with the goal of making subtypes of this base hazard easily customizable using variables. The goal is to make it as easy as possible for mappers to make and customize hazards based on the subtypes listed below.

### **Hazard Basics**
Most hazards are activated randomly or on contact. Random hazard effects will constantly repeat with a delay set between random_min and random_max. Contact hazard effects either happen when walked over (not dense) or bumped (dense), with a short cooldown. As of now, only radiation barrels use the attacked hazard trigger, which happens when the hazard is, of course, attacked.


### **Hazard Subtypes**
base hazard
all hazards can choose to add slowdown
hazard/atmospheric
a simple hazard that emits a chosen gas, with set mols and temp
hazard/electrical
hazards that emit sparks, stuns, or tesla arcs. The bread and butter of this system
hazard/floor
floor hazards, such as spikes, slips, and launchers
hazard/radioactive
the old radioactive barrels, with their activation procs moved to the base hazard. otherwise nearly the same!
hazard/slowdown
require climbing over, under, or have a stick chance
hazard/spray
sprays foam, smoke, or water.

implemented electrical hazard example:
https://youtu.be/4P-67V_US6Q

hazard example suite as of july 22. more are likely to come as I get more recommendations.
https://youtu.be/gLLdpiM-Ofg

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Makes it (hopefully) easy for mappers to customize simple hazards to their needs, and adds new kinds of hazards that don't rely on sim mechanics nearly as heavily. This allows for hazards to be easily used in ruins without much consideration, compared to something like emitters.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

:cl:
add: adds ruin hazards
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Thera-Pissed committed:
basic stuff

thgvr reviewed (commented):
https://github.com/shiptest-ss13/Shiptest/blob/master/.github/AUTODOC_GUIDE.md

Thera-Pissed committed:
autodoc and much better comments

Thera-Pissed committed:
living/carbon to appease apply_damage

Thera-Pissed committed:
temp notes file is failing linters, wee

Thera-Pissed committed:
atmos, foam, chem smoke, spray, floor rework.

Thera-Pissed committed:
linter dumb

Thera-Pissed committed:
linter again

Thera-Pissed committed:
is this linters bad ? help

Thera-Pissed committed:
helmcrab moving code around fails again

Thera-Pissed committed:
Merge branch 'master' into electrical-hazards

Thera-Pissed committed:
moves nearby check from hazard/radioactive to base hazard

this helps clean up making radioactive a hazard subtype, though it could be cleaner. Should tweak barrels to not be the base hazard, and add the radiation triggers to other hazards.

Thera-Pissed committed:
adds client_range var because uh, choosing stuff is good?

Thera-Pissed committed:
cleaner processing code

Thera-Pissed committed:
moves process logic to do_random_effect repeated

hey wait a second, doing something once every cooldown? I already do that with random_effect. why also need this thing to process when I could just use the existing setup! yeah

Thera-Pissed committed:
powered shutoff

Powered shutoff based on the medipen vault console, which requires a cable connection to charge, and at max allows the shutoff to be used. Could be more tactile but this works for now.

Thera-Pissed committed:
example powered

thgvr commented:
Still being worked on? Let me know when you want a review

Thera-Pissed commented:
still being worked on, been busy this week, mostly drafted because I want more implemented hazards before its sent

Thera-Pissed commented:
if the electric hazards are wanted let me know, I'm not finishing the rest of this however.

Thera-Pissed committed:
comment clarification

Looking at the code again, hopefully these comments are more clear.

Thera-Pissed committed:
Merge branch 'master' into electrical-hazards

Thera-Pissed committed:
unused code

Thera-Pissed committed:
merge fixes

thgvr commented:
Let us know when you want a review on this

Thera-Pissed committed:
Merge remote-tracking branch 'upstream/master' into electrical-hazards

Thera-Pissed commented:
My goal of having several hazards finished for this before review has made me stall on it for a while, if I could get a review now that'd be great. If anything seems wonky I'll cut it for a future PR.

FalloutFalcon commented on code/modules/ruins/objects_and_mobs/generic_hazards/hazards/electrical_conduit.dm:
This file is not ticked

FalloutFalcon commented on code/modules/ruins/objects_and_mobs/generic_hazards/hazard_code/floor.dm:
```diff
@@ -0,0 +1,129 @@
+/obj/structure/hazard/floor/proc/launch_sequence()
+	visible_message(span_warning(launch_warning))
+	icon_state = initial(icon_state) + "-launch"
+	sleep(10)
```
Try and avoid using sleep as its prone to causing overtime, I would swap this to an addtimer with a callback instead
something like 
```
addtimer(CALLBACK(src, PROC_REF(acctually_launch)), 0.5 SECONDS)
```

FalloutFalcon commented:
ehh this has been open for 2 long.. just remeber in future that sleeps are evil..
